### PR TITLE
Adding files for directory structure

### PR DESCRIPTION
**Pull request by anandmohan777**

### CODE REVIEW

File: src/test/java/uk/ac/ebi/ega/egafuse/runner/CommandLineOptionParserTest.java [hunk 1]
```diff
@@ -107,4 +109,22 @@ public void parser_WhenGivenMountPathNotExists_ThenThrowsException() throws IOEx
...
+        assertFalse(cliConfigurationValues.isTreeStructureEnable());
+    }
+    
+    @Test
+    public void parser_WhenGivenTreeOptionEnable_ThenReturnsTrue() throws IOException{        
+        final File mountFolder = temporaryFolder.newFolder("tmp", "mount");
+        String[] args = { "-m", mountFolder.toPath().toAbsolutePath().toString(), "-u", "amohan", "-p", "testpass", "-t", "enable"};
+        OptionSet set = CommandLineOptionParser.buildParser().parse(args) ;        
+        CliConfigurationValues cliConfigurationValues = CommandLineOptionParser.parser(set);
+        assertTrue(cliConfigurationValues.isTreeStructureEnable());
```
Comment (afoix) [inline src/test/java/uk/ac/ebi/ega/egafuse/runner/CommandLineOptionParserTest.java:140]: Do we test what happens if they give us a wrong input in the -t parameter?

File: src/test/resources/application-test.properties [hunk 2]
```diff
@@ -7,5 +7,6 @@ connection.maxAttempts=6
 connection.backoff=10000
 cache.prefetch=1
 api.chunksize=10485760
+tree=false
```
Comment (afoix) [inline src/test/resources/application-test.properties:10]: Do this means that the default value for tree is false?

File: src/test/java/uk/ac/ebi/ega/egafuse/runner/CommandLineOptionParserTest.java (hunk 1 above)
Comment (anandmohan777) [reply src/test/java/uk/ac/ebi/ega/egafuse/runner/CommandLineOptionParserTest.java:140]: Added a new test case now for this condition, i.e, invalid input in -t, it will discard that and defaults it to enable.

File: src/test/resources/application-test.properties (hunk 2 above)
Comment (anandmohan777) [reply src/test/resources/application-test.properties:10]: For only test cases defaults tree is disabled but I have written test cases for both enable and disable

File: README.md [hunk 3]
```diff
@@ -68,6 +68,7 @@ Optional arguments:
 * m : mount point path, default value: /tmp/mnt `Note: Ensure that the mount point path exists`
 * cache : the maximum size of the cache, default value: 100 `Means 100 * 10 MB = 1000 MB`
 * c : connections, download using specified number of connections, default value: 4
+* t : tree directory structure, show file in proper directory structure, allowed values: disable & enable, default value: `enable`
```
Comment (jorizci) [inline README.md]: I would rephrase it as: Toggle tree structure Shows files with original submitter directory structure (enable) or all the dataset files in a single directory (disable), default value: `enable`

**Review by jorizci — CHANGES_REQUESTED** [review]
In general is ok, small thing on the comments, I would test more different cases for the correct parsing of the tree something like

/a/test.cip
/a/test2.cip
/a/b/test.cip
/test.cip

And checking that all the files and structure is correct :)

File: src/main/java/uk/ac/ebi/ega/egafuse/config/EgaFuseApplicationConfig.java [hunk 4]
```diff
@@ -53,6 +53,13 @@
 
     @Value("${api.chunksize}")
     private long chunkSize;
+    
+    public static boolean isTreeStructureEnable;
+    
+    @Value("${tree}")
+    public void setFileDatabaseService(boolean isTreeStructure) {
+        isTreeStructureEnable = isTreeStructure;
+    }
```
Comment (gariem) [inline src/main/java/uk/ac/ebi/ega/egafuse/config/EgaFuseApplicationConfig.java:62]: Not sure if this can be skipped and @Value("${api. tree}") can be directly injected to CliConfigurationValues class?

File: src/main/java/uk/ac/ebi/ega/egafuse/runner/CommandLineOptionPropertySource.java [hunk 5]
```diff
@@ -46,6 +46,8 @@ public Object getProperty(String name) {
             return String.valueOf(cliConfigurationValues.getConnection());
         case "mountPath":
             return cliConfigurationValues.getMountPath().toString();
+        case "tree":
+            return cliConfigurationValues.isTreeStructureEnable();
```
Comment (gariem) [inline src/main/java/uk/ac/ebi/ega/egafuse/runner/CommandLineOptionPropertySource.java:50]: I think it's a good time to refactor this and use constants or enums

File: src/main/java/uk/ac/ebi/ega/egafuse/service/EgaDirectory.java [hunk 6]
```diff
@@ -83,7 +85,12 @@ public synchronized void read(Pointer buf, FuseFillDir filler) {
             if (getName().equalsIgnoreCase("datasets")) {
                 egaDatasetService.getDatasets().forEach(this::add);
             } else {
-                egaFileService.getFiles(this).forEach(this::add);
+                List<EgaFile> egaFiles = egaFileService.getFiles(this);
+
+                if (isTreeStructureEnable)
+                    egaDatasetService.buildSubDirectoryFromFilePath(egaFiles, this);
+                else
+                    egaFiles.stream().forEach(this::add);
```
Comment (gariem) [inline src/main/java/uk/ac/ebi/ega/egafuse/service/EgaDirectory.java:93]: I think overriding the add method or creating a new one addAll would be more convenient than calling the single add many times

File: src/main/java/uk/ac/ebi/ega/egafuse/service/EgaDirectory.java [hunk 7]
```diff
@@ -83,7 +85,12 @@ public synchronized void read(Pointer buf, FuseFillDir filler) {
             if (getName().equalsIgnoreCase("datasets")) {
```
Comment (gariem) [inline src/main/java/uk/ac/ebi/ega/egafuse/service/EgaDirectory.java]: This is generally a risky way to use the equals methods, I think we should take the chance to make it NPE safe

File: src/main/java/uk/ac/ebi/ega/egafuse/runner/CommandLineOptionParser.java [hunk 8]
```diff
@@ -62,6 +62,9 @@ public static CliConfigurationValues parser(OptionSet optionSet) throws IOExcept
                     .concat(" can't be used as mount point. Ensure that the directory path should exist."));
         }
         cliConfigurationValues.setMountPath(mntPath);
+        cliConfigurationValues
+                .setTreeStructureEnable(optionSet.valueOf("t").toString().equalsIgnoreCase("disable") ? false : true);
```
Comment (gariem) [inline src/main/java/uk/ac/ebi/ega/egafuse/runner/CommandLineOptionParser.java]: Again this usage of the equals method is not null pointer exception safe

**Review by gariem — CHANGES_REQUESTED** [review]

File: src/main/java/uk/ac/ebi/ega/egafuse/runner/CommandLineOptionParser.java (hunk 8 above)
Comment (anandmohan777) [reply src/main/java/uk/ac/ebi/ega/egafuse/runner/CommandLineOptionParser.java]: Set the default value to enable in optionSet.valueOf("t") so NPE will never occurr

File: src/main/java/uk/ac/ebi/ega/egafuse/service/EgaDirectory.java (hunk 6 above)
Comment (anandmohan777) [reply src/main/java/uk/ac/ebi/ega/egafuse/service/EgaDirectory.java:93]: We are also setting the parent value for each file in the add method. So can't use addall, we would still need to loop around each file and add it.

File: src/main/java/uk/ac/ebi/ega/egafuse/config/EgaFuseApplicationConfig.java (hunk 4 above)
Comment (anandmohan777) [reply src/main/java/uk/ac/ebi/ega/egafuse/config/EgaFuseApplicationConfig.java:62]: It is used in EgaDirectory class and we can inject values directly in it, however, there are a lot of places from where  EgaDirectory is created so it is difficult to inject this in all those classes. So I decided to make it to use as a constant

**Comment by anandmohan777** [comment]
@jorizci added the test case for that scenario having different file and folder structure you have mentioned.

**Review by jorizci — CHANGES_REQUESTED** [review]
Sorry for being a pain, could you do a test where the calls to the rest services are mocked and we see it generating the structure correctly. As far as you have proved, you can use those methods and look ok but there is no case were we can see that for X response in the json message this directory structure is generated.

**Review by gariem — APPROVED** [review]

**Review by afoix — APPROVED** [review]

**Review by jorizci — APPROVED** [review]
